sdk/python/feast/infra/offline_stores/third_party/spark.py

```diff
@@ -0,0 +1,529 @@
+            spark_session=spark_session,
+            table_name=tmp_entity_df_table_name,
+            entity_df=entity_df,
+        )
+        event_timestamp_col = offline_utils.infer_event_timestamp_from_entity_df(
+            entity_schema=entity_schema,
+        )
+        entity_df_event_timestamp_range = _get_entity_df_event_timestamp_range(
+            entity_df, event_timestamp_col, spark_session,
+        )
+
+        expected_join_keys = offline_utils.get_expected_join_keys(
+            project=project, feature_views=feature_views, registry=registry
+        )
+        offline_utils.assert_expected_columns_in_entity_df(
+            entity_schema=entity_schema,
+            join_keys=expected_join_keys,
+            entity_df_event_timestamp_col=event_timestamp_col,
+        )
+
+        query_context = offline_utils.get_feature_view_query_context(
+            feature_refs,
+            feature_views,
+            registry,
+            project,
+            entity_df_event_timestamp_range,
+        )
+
+        query = offline_utils.build_point_in_time_query(
+            feature_view_query_contexts=query_context,
+            left_table_query_string=tmp_entity_df_table_name,
+            entity_df_event_timestamp_col=event_timestamp_col,
+            entity_df_columns=entity_schema.keys(),
+            query_template=MULTIPLE_FEATURE_VIEW_POINT_IN_TIME_JOIN,
+            full_feature_names=full_feature_names,
+        )
+
+        # TODO: Figure out what this is used for
+        # on_demand_feature_views = OnDemandFeatureView.get_requested_odfvs(
+        #     feature_refs=feature_refs, project=project, registry=registry
+        # )
+
+        return SparkRetrievalJob(
+            spark_session=spark_session,
+            query=query,
+            full_feature_names=full_feature_names,
+            on_demand_feature_views=OnDemandFeatureView.get_requested_odfvs(
+                feature_refs, project, registry
+            ),
+            metadata=RetrievalMetadata(
+                features=feature_refs,
+                keys=list(set(entity_schema.keys()) - {event_timestamp_col}),
+                min_event_timestamp=entity_df_event_timestamp_range[0],
+                max_event_timestamp=entity_df_event_timestamp_range[1],
+            ),
+        )
+
+    @staticmethod
+    def pull_all_from_table_or_query(
+        config: RepoConfig,
+        data_source: DataSource,
+        join_key_columns: List[str],
+        feature_name_columns: List[str],
+        event_timestamp_column: str,
+        start_date: datetime,
+        end_date: datetime,
+    ) -> RetrievalJob:
+        """
+        Note that join_key_columns, feature_name_columns, event_timestamp_column, and created_timestamp_column
+        have all already been mapped to column names of the source table and those column names are the values passed
+        into this function.
+        """
+        return SparkOfflineStore.pull_latest_from_table_or_query(
```

Comment:
This seems weird? Other offline stores create a different query that generates all feature values between a time range. Any reason why this is implemented differently?